app/controllers/devise_sessions_controller.rb

```diff
@@ -3,54 +3,81 @@ class DeviseSessionsController < Devise::SessionsController
   include CookiesHelper
 
   before_action :check_login_state, only: %i[
-    create
     phone_code
-    phone_code_send
     phone_verify
     phone_resend
   ]
 
   before_action :check_password_ok, only: %i[
     phone_code
-    phone_code_send
     phone_verify
     phone_resend
   ]
 
   def create
-    render :new and return unless params.dig(:user, :password)
+    payload = ApplicationKey.validate_jwt!(params[:jwt]) if params[:jwt]
+
+    render :new and return unless params.dig(:user, :email) || params.dig(:user, :password)
+
+    @email = params.dig(:user, :email)
+    catch(:warden) do
+      self.resource = warden.authenticate(auth_options)
+    end
 
-    self.resource = warden.authenticate(auth_options)
     if resource
-      login_state.update!(password_ok: true)
+      @login_state = create_login_state(payload, @email, password_ok: true)
+      @login_state_id = login_state.id
+      session[:login_state_id] = login_state.id
+
       if request.env["warden.mfa.required"]
         MultiFactorAuth.generate_and_send_code(resource)
         redirect_to user_session_phone_code_path
       else
         do_sign_in
       end
     else
-      @password_error_message =
-        case User.find_by(email: params.dig(:user, :email))&.unauthenticated_message
-        when :last_attempt
-          I18n.t("devise.failure.last_attempt")
-        when :locked
-          I18n.t("devise.failure.locked")
-        else
-          I18n.t("devise.sessions.new.fields.password.errors.incorrect")
-        end
-      render :new
-    end
-  end
+      @resource_error_messages = {}
 
-  def phone_code; end
+      @resource_error_messages[:email] = if @email.blank?
+                                           [I18n.t("activerecord.errors.models.user.attributes.email.blank")]
+                                         elsif !Devise.email_regexp.match? @email
+                                           [I18n.t("activerecord.errors.models.user.attributes.email.invalid")]
+                                         end
+
+      if params.dig(:user, :password).blank?
+        @resource_error_messages[:password] = [I18n.t("activerecord.errors.models.user.attributes.password.blank")]
+      end
 
-  def phone_code_send
```

Comment:
The button on https://www.account.staging.publishing.service.gov.uk/sign-in/phone/resend now leads to a 404

Reply:
Ah, same as in the registration controller, I'll add this to the `phone_resend` method.